app/ide-desktop/lib/cloud/src/authentication/src/authentication.tsx

```diff
@@ -0,0 +1,644 @@
+/**
+ * @file Module for authenticating users with AWS Cognito.
+ * 
+ * Provides an `AuthProvider` component that wraps the entire application, and a `useAuth` hook that
+ * can be used from any React component to access the currently logged-in user's session data. The
+ * hook also provides methods for registering a user, logging in, logging out, etc.
+ */
+import { ComponentType, createContext, FC, ReactElement, ReactNode, useCallback, useContext, useEffect, useMemo, useState } from 'react';
+import { Navigate, Outlet, useNavigate, useOutletContext } from 'react-router-dom';
+import { toast } from "react-hot-toast"
+
+import { getUsersMe, Organization, SetUsernameBody } from './api';
+import * as api from './api';
+import authApi, { isAmplifyError, isAuthError, AuthConfig, Api } from './authentication/api';
+import { DASHBOARD_PATH, Logger, LOGIN_PATH, REGISTRATION_PATH, RESET_PASSWORD_PATH, SET_USERNAME_PATH } from './components/app';
+import { useAsyncEffect } from './hooks';
+
+
+
+// =============
+// === Types ===
+// =============
+
+
+// === UserSession ===
+
+export type UserSession =
+  | FullUserSession
+  | PartialUserSession;
+
+/// The type of the object containing the currently signed-in user's session data.
+export interface FullUserSession {
+  /// Literal for narrowing the type of the `Session` object.
+  state: "full",
+  /// The user's JSON Web Token (JWT), used for authenticating and authorizing requests to the API.
+  accessToken: string;
+  /// The user's email address.
+  email: string;
+  /// The user's organization information.
+  organization: Organization;
+}
+
+/// The type of the object containing the currently signed-in user's session data, if the user has
+/// not yet set their username.
+///
+/// If a user has not yet set their username, they do not yet have an organization associated with
+/// their account. Otherwise, this type is identical to the `Session` type. This type should ONLY be
+/// used by the `SetUsername` component.
+export interface PartialUserSession {
+  /// Literal for narrowing the type of the `Session` object.
+  state: "partial",
+  /// The user's JSON Web Token (JWT), used for authenticating and authorizing requests to the API.
+  accessToken: string;
+  /// The user's email address.
+  email: string;
+}
+
+
+// =================
+// === Constants ===
+// =================
+
+const SIGN_UP_SUCCESS = "We have sent you an email with further instructions!"
+const CONFIRM_SIGN_UP_SUCCESS = "Your account has been confirmed! Please log in."
+const SET_USERNAME_SUCCESS = "Your username has been set!"
+const SIGN_IN_WITH_PASSWORD_SUCCESS = "Successfully logged in!"
+const FORGOT_PASSWORD_SUCCESS = "We have sent you an email with further instructions!"
+const RESET_PASSWORD_SUCCESS = "Successfully reset password!"
+const SIGN_OUT_SUCCESS = "Successfully logged out!"
```

Comment:
1. constants should be defined on top of the file.
2. Constants should be grouped - these are messages. Why not doing it this way instead?:

```
const messages = {
    sign_up_success: ...
    ...
}
```